integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py

```diff
@@ -114,26 +118,31 @@ def run(self, messages: List[ChatMessage], generation_kwargs: Optional[Dict[str,
         formatted_messages = [_convert_message_to_llamacpp_format(msg) for msg in messages]
 
         response = self.model.create_chat_completion(messages=formatted_messages, **updated_generation_kwargs)
-        replies = [
-            ChatMessage(
-                content=choice["message"]["content"],
-                role=ChatRole[choice["message"]["role"].upper()],
-                name=None,
-                meta={
-                    "response_id": response["id"],
-                    "model": response["model"],
-                    "created": response["created"],
-                    "index": choice["index"],
-                    "finish_reason": choice["finish_reason"],
-                    "usage": response["usage"],
-                },
-            )
-            for choice in response["choices"]
-        ]
-
-        for reply, choice in zip(replies, response["choices"]):
+
+        replies = []
+
+        for choice in response["choices"]:
+            meta = {
+                "response_id": response["id"],
+                "model": response["model"],
+                "created": response["created"],
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
+                "usage": response["usage"],
+            }
+
+            name = None
             tool_calls = choice.get("message", {}).get("tool_calls", [])
             if tool_calls:
-                reply.meta["tool_calls"] = tool_calls
-        reply.name = tool_calls[0]["function"]["name"] if tool_calls else None
+                meta["tool_calls"] = tool_calls
+                name = tool_calls[0]["function"]["name"]
+
+            reply = ChatMessage.from_assistant(choice["message"]["content"], meta=meta)
+            if name:
+                if hasattr(reply, "_name"):
+                    reply._name = name  # new ChatMessage
+                elif hasattr(reply, "name"):
+                    reply.name = name  # legacy ChatMessage
```

Comment:
once we release 2.9.0, we can get rid of this ugly check

Reply:
👍  after 2.9.0 release, we may consider to refactor `for`'s code in a `_create_reply` method maybe.